src/dmd/inline.d

```diff
@@ -226,25 +216,54 @@ public:
     }
 
     override void visit(IfStatement s)
+    {
+        visitIfStatement(s, null);
+    }
+
+    extern (D) private void visitIfStatement(IfStatement s, void delegate(DoInlineAs!Result) visitSubseq)
     {
         assert(!s.prm);
         auto econd = doInlineAs!Expression(s.condition, ids);
         assert(econd);
 
         auto ifbody = doInlineAs!Result(s.ifbody, ids);
-        bool bodyReturn = ids.foundReturn;
-
+        bool thenReturn = ids.foundReturn;
         ids.foundReturn = false;
+
         auto elsebody = doInlineAs!Result(s.elsebody, ids);
+        bool elseReturn = ids.foundReturn;
+        ids.foundReturn = false;
 
         static if (asStatements)
         {
+            ids.foundReturn = thenReturn && elseReturn;
             result = new IfStatement(s.loc, s.prm, econd, ifbody, elsebody, s.endloc);
         }
         else
         {
             alias e1 = ifbody;
             alias e2 = elsebody;
+
+            if (thenReturn != elseReturn && visitSubseq)
+            {
+                scope DoInlineAs!Result v = new DoInlineAs!Result(ids);
+                visitSubseq(v);
+                if (!thenReturn)
+                {
+                    e1 = Expression.combine(e1, v.result);
+                    thenReturn = v.ids.foundReturn;
```

Comment:
These rewrites are pretty important, and need a test case to cover them.

Reply:
Test I pasted covers this too.